core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -579,9 +579,33 @@ class UnifiedLog(@volatile var logStartOffset: Long,
     result
   }
 
-  def hasOngoingTransaction(producerId: Long): Boolean = lock synchronized {
-    val entry = producerStateManager.activeProducers.get(producerId)
-    entry != null && entry.currentTxnFirstOffset.isPresent
+  def transactionNeedsVerifying(producerId: Long, producerEpoch: Short, baseSequence: Int): Boolean = lock synchronized {
+    val entry = producerStateManager.entryForVerification(producerId, producerEpoch, baseSequence)
+    (!entry.currentTxnFirstOffset.isPresent) &&
+      (entry.compareAndSetVerificationState(producerEpoch, ProducerStateEntry.VerificationState.EMPTY, ProducerStateEntry.VerificationState.VERIFYING) ||
+        entry.verificationState() == ProducerStateEntry.VerificationState.VERIFYING)
+  }
+  
+  def compareAndSetVerificationState(producerId: Long,
+                                     producerEpoch: Short,
+                                     baseSequence: Int,
+                                     expectedVerificationState: ProducerStateEntry.VerificationState,
+                                     newVerificationState: ProducerStateEntry.VerificationState): Unit = { lock synchronized {
```

Comment:
nit: can drop braces around `lock synchronized`?

Reply:
That's the method braces. Should I just make a new line so it is clearer?